Tools & Scripts that build testapps

sunmou99 opened the pull request:
Tested locally with Unity `2017.4.37f1` and  `2019.2.8f1`. 
`firestore` testapp uses [Unity Test Framework](https://docs.unity3d.com/Packages/com.unity.test-framework@1.1/manual/index.html), which requires at least Unity 2019.2. 
Except for firestore in `2017.4.37f1` and iOS analytics testapps, the rest testapps build successfully with All platform: `iOS,Android,Win64,OSXUniversal,Linux64`.
For in-editor test, MacOS' Gatekeeper (which blocks untrusted binaries) will prevent `playmode` tests on MacOS.
- Migrated Tools that automate the test. Tools are `MenuScene`, `automated_testapp` and etc., under path`scripts/gha/integration_test` 
- Migrated Scripts that build testapps. Python scripts under path `scripts/gha`. Modified the scripts and configs for GitHub.
- Updated the Unity Project setting. 

To test locally:
1. Prerequisite:
Make sure you have Unity 2017.4.37f1 installed and activated. 
For iOS, using XCode 10 by default,
For Android, using SDK tool 25.2.*, JAVA_HOME set to JDK8
(For Unity 2019.2.8f1, we need newer JDK, and Android SDK)
2. Restore secret: `python scripts/gha/restore_secrets.py --passphrase ${passphrase}`
3. Build testapps: `python scripts/gha/build_testapps.py --t auth,database --u 2017.4.37f1 --p iOS,Android,Win64,OSXUniversal,Linux64 --force_latest_runtime`

a-maurice commented on analytics/testapp/ProjectSettings/ProjectSettings.asset:
```diff
@@ -166,7 +166,7 @@ PlayerSettings:
     iOS: 0
   AndroidBundleVersionCode: 1
-  AndroidMinSdkVersion: 16
+  AndroidMinSdkVersion: 26
```
I'm not a fan of this being updated.  We should figure out what the issue here is, because 16 should be fine, and bumping this up just hides the problem.

sunmou99 replied:
Update the version for now to unblock the build process.
Shawn created a ticket for this problem, and VJ is working on it. Will revert the change once it problem solved.

sunmou99 replied:
We have a ticket. And talked with Shawn and VJ today, we are working on it now.

a-maurice reviewed (commented):
This is really big, and thus hard to actually review.  I know most of this is being brought over from the internal versions of these files.  Can you call out the major changes that had to be done, cause that will make it easier to note what is changing.

sunmou99 commented:
> This is really big, and thus hard to actually review. I know most of this is being brought over from the internal versions of these files. Can you call out the major changes that had to be done, cause that will make it easier to note what is changing.

`build_testapps.py`: [copy](https://github.com/firebase/firebase-unity-sdk/pull/41/files#diff-3c7ca8ab01f93a7ee4419bc47f3a7dfa3e9d81f1db31e1cd4a0238912992f5d4R367) tools that automate the test to our unity project.
`build_testapps.json`: updated app info (name & path), and deleted obsolete configs.
`xcodebuild.py`: build for unsigned iOS apps.

yifanyang commented on scripts/gha/integration_testing/xcodebuild.py:
```diff
@@ -0,0 +1,114 @@
+  """
+  iphone_build_dir = os.path.join(output_dir, configuration + "-iphoneos")
+  payload_path = os.path.join(iphone_build_dir, "Payload")
+  for _, directories, _ in os.walk(iphone_build_dir):
+    for directory in directories:
+      if directory.endswith('.app'):
```
Would it be simpler if we use [`glob`](https://docs.python.org/3/library/glob.html) here?

sunmou99 replied:
Good to know! We use `os.walk` in a few other scripts. Maybe worthing have a separate PR solve them at once.

yifanyang commented on scripts/gha/integration_testing/xcodebuild.py:
```diff
@@ -0,0 +1,114 @@
+        shutil.move(app_path, payload_path)
+        shutil.make_archive(
+            payload_path, "zip", root_dir=iphone_build_dir, base_dir="Payload")
+        shutil.move("%s.%s"%(payload_path, "zip"), ipa_path)
```
[f-string](https://docs.python.org/3/reference/lexical_analysis.html#formatted-string-literals) appears to be the modern way for formatting strings. I think they are more readable than `str.format()` and `%` operator. It's up to you.

sunmou99 replied:
`f-string` is a new feature in python 3.6. Our script used to support python 2. But I think that it's worth doing this cleaning for the python scripts.

yifanyang reviewed (approved):
I skimmed through all the python code and left a few comments, but they look largely good to me. I hope that once I'm more familiar with the Games stacks, I'll be able to review more type of code.

yifanyang commented:
> > This is really big, and thus hard to actually review. I know most of this is being brought over from the internal versions of these files. Can you call out the major changes that had to be done, cause that will make it easier to note what is changing.
> 
> `build_testapps.py`: [copy](https://github.com/firebase/firebase-unity-sdk/pull/41/files#diff-3c7ca8ab01f93a7ee4419bc47f3a7dfa3e9d81f1db31e1cd4a0238912992f5d4R367) tools that automate the test to our unity project.
> `build_testapps.json`: updated app info (name & path), and deleted obsolete configs.
> `xcodebuild.py`: build for unsigned iOS apps.

I feel it will be clearer if this code change is structured as two separate pull requests:
- one PR that consists of pure copy-pasting old code from google3
- another PR that contains the real change that needs in-depth review

This would make the PR easier to review and faster to merge ([go/small-cls](http://go/small-cls)). It's okay to me if it's hard to split this time, but maybe starting from next PR, let's be more conscious of the size of our PR. What do you think?

cynthiajoan commented on scripts/gha/build_testapps.py:
```diff
@@ -0,0 +1,968 @@
+
+
+def _add_automated_test_runner(dir_helper):
+  """Copies automated_testpp."""
```
nit: testapp

cynthiajoan commented on scripts/gha/integration_testing/MenuScene/SceneLoader.cs:
```diff
@@ -0,0 +1,51 @@
+using System;
```
Add copyright info

cynthiajoan commented on scripts/gha/integration_testing/MenuScene/SceneLoader.cs:
```diff
@@ -0,0 +1,51 @@
+
+  // After loading this scene, the automated scene will be loaded by default
+  // after this many seconds.
+  const float TIMEOUT = 6f;
```
Can you add some comments about why 6?

cynthiajoan reviewed (approved):
A few nit comments, this is really an epic PR, great job!
